File: collect_app/src/androidTest/java/org/odk/collect/android/support/TestScheduler.java
package org.odk.collect.android.support;

import android.content.Context;

import androidx.test.core.app.ApplicationProvider;
import androidx.work.WorkManager;

import org.jetbrains.annotations.NotNull;
import org.odk.collect.async.Cancellable;
import org.odk.collect.async.CoroutineAndWorkManagerScheduler;
import org.odk.collect.async.Scheduler;
import org.odk.collect.async.TaskSpec;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Supplier;

public class TestScheduler implements Scheduler {

    private final Scheduler wrappedScheduler;

    private final Object lock = new Object();
    private int tasks;
    private Runnable finishedCallback;

    private final List<DeferredTask> deferredTasks = new ArrayList<>();

    public TestScheduler() {
        WorkManager workManager = WorkManager.getInstance(ApplicationProvider.getApplicationContext());
        this.wrappedScheduler = new CoroutineAndWorkManagerScheduler(workManager);
    }

    @Override
    public Cancellable repeat(@NotNull Runnable foreground, long repeatPeriod) {
        increment();

        return wrappedScheduler.repeat(() -> {
            foreground.run();
            decrement();
        }, repeatPeriod);
    }

    @Override
    public <T> void immediate(@NotNull Supplier<T> foreground, @NotNull Consumer<T> background) {
        increment();

        wrappedScheduler.immediate(foreground, t -> {
            background.accept(t);
            decrement();
        });
    }

    @Override
    public void immediate(@NotNull Runnable foreground) {
        increment();

        wrappedScheduler.immediate(() -> {
            foreground.run();
            decrement();
        });
    }

    @Override
    public void networkDeferred(@NotNull String tag, @NotNull TaskSpec spec, @NotNull Map<String, String> inputData) {
        deferredTasks.add(new DeferredTask(tag, spec, null, inputData));
    }

    @Override
    public void networkDeferred(@NotNull String tag, @NotNull TaskSpec spec, long repeatPeriod, @NotNull Map<String, String> inputData) {
        cancelDeferred(tag);
        deferredTasks.add(new DeferredTask(tag, spec, repeatPeriod, inputData));
    }

    @Override
    public void cancelDeferred(@NotNull String tag) {
        deferredTasks.removeIf(t -> t.getTag().equals(tag));
    }

    @Override
    public boolean isDeferredRunning(@NotNull String tag) {
        return wrappedScheduler.isDeferredRunning(tag);
    }

    public void runDeferredTasks() {
        Context applicationContext = ApplicationProvider.getApplicationContext();

        for (DeferredTask deferredTask : deferredTasks) {
            deferredTask.getSpec().getTask(applicationContext, deferredTask.getInputData()).get();
        }

        // Remove non repeating tasks
        deferredTasks.removeIf(deferredTask -> deferredTask.repeatPeriod == null);
    }

    public void setFinishedCallback(Runnable callback) {
        this.finishedCallback = callback;
    }

    private void increment() {
        synchronized (lock) {
            tasks++;
        }
    }

    private void decrement() {
        synchronized (lock) {
            tasks--;

            if (tasks == 0 && finishedCallback != null) {
                finishedCallback.run();
            }
        }
    }

    public int getTaskCount() {
        synchronized (lock) {
            return tasks;
        }
    }

    public List<DeferredTask> getDeferredTasks() {
        return deferredTasks;
    }

    @Override
    public void cancelAllDeferred() {
    }

    public static class DeferredTask {

        private final String tag;
        private final TaskSpec spec;
        private final Long repeatPeriod;
        private final Map<String, String> inputData;

        public DeferredTask(String tag, TaskSpec spec, Long repeatPeriod, Map<String, String> inputData) {
            this.tag = tag;
            this.spec = spec;
            this.repeatPeriod = repeatPeriod;
            this.inputData = inputData;
        }

        public String getTag() {
            return tag;
        }

        public TaskSpec getSpec() {
            return spec;
        }

        public long getRepeatPeriod() {
            return repeatPeriod;
        }

        public Map<String, String> getInputData() {
            return inputData;
        }
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/support/TestScheduler.kt
package org.odk.collect.android.support

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.work.WorkManager
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Runnable
import kotlinx.coroutines.flow.Flow
import org.odk.collect.async.Cancellable
import org.odk.collect.async.CoroutineAndWorkManagerScheduler
import org.odk.collect.async.NotificationInfo
import org.odk.collect.async.Scheduler
import org.odk.collect.async.TaskSpec
import org.odk.collect.async.network.NetworkStateProvider
import java.util.function.Consumer
import java.util.function.Supplier
import kotlin.coroutines.CoroutineContext

class TestScheduler(private val networkStateProvider: NetworkStateProvider) : Scheduler, CoroutineDispatcher() {

    private val wrappedScheduler: Scheduler
    private val lock = Any()
    private var tasks = 0
    private var finishedCallback: Runnable? = null
    private val deferredTasks: MutableList<DeferredTask> = ArrayList()
    private val backgroundDispatcher = Dispatchers.IO

    init {
        val workManager = WorkManager.getInstance(ApplicationProvider.getApplicationContext())
        wrappedScheduler = CoroutineAndWorkManagerScheduler(Dispatchers.Main, this, workManager)
    }

    override fun repeat(foreground: Runnable, repeatPeriod: Long): Cancellable {
        return wrappedScheduler.repeat({ foreground.run() }, repeatPeriod)
    }

    override fun <T> immediate(background: Supplier<T>, foreground: Consumer<T>) {
        increment()
        wrappedScheduler.immediate(background) { t: T ->
            foreground.accept(t)
            decrement()
        }
    }

    override fun immediate(foreground: Boolean, delay: Long?, runnable: Runnable) {
        increment()
        wrappedScheduler.immediate(foreground, delay) {
            runnable.run()
            decrement()
        }
    }

    override fun immediate(
        tag: String,
        spec: TaskSpec,
        inputData: Map<String, String>,
        notificationInfo: NotificationInfo
    ) {
        increment()
        val context = ApplicationProvider.getApplicationContext<Context>()
        wrappedScheduler.immediate {
            spec.getTask(context, inputData, true).get()
            decrement()
        }
    }

    override fun networkDeferred(
        tag: String,
        spec: TaskSpec,
        inputData: Map<String, String>,
        networkConstraint: Scheduler.NetworkType?
    ) {
        cancelDeferred(tag)
        deferredTasks.add(DeferredTask(tag, spec, null, inputData, networkConstraint))
    }

    override fun networkDeferredRepeat(
        tag: String,
        spec: TaskSpec,
        repeatPeriod: Long,
        inputData: Map<String, String>
    ) {
        cancelDeferred(tag)
        deferredTasks.add(DeferredTask(tag, spec, repeatPeriod, inputData, null))
    }

    override fun cancelDeferred(tag: String) {
        deferredTasks.removeIf { t: DeferredTask -> t.tag == tag }
    }

    override fun isDeferredRunning(tag: String): Boolean {
        return wrappedScheduler.isDeferredRunning(tag)
    }

    fun runDeferredTasks() {
        if (networkStateProvider.isDeviceOnline) {
            val applicationContext = ApplicationProvider.getApplicationContext<Context>()
            deferredTasks.removeIf { deferredTask ->
                if (deferredTask.networkConstraint == null || deferredTask.networkConstraint == networkStateProvider.currentNetwork) {
                    deferredTask.spec.getTask(applicationContext, deferredTask.inputData, true)
                        .get()
                    deferredTask.repeatPeriod == null
                } else {
                    false
                }
            }
        }
    }

    fun setFinishedCallback(callback: Runnable?) {
        finishedCallback = callback
    }

    private fun increment() {
        synchronized(lock) { tasks++ }
    }

    private fun decrement() {
        synchronized(lock) {
            tasks--
            if (tasks == 0 && finishedCallback != null) {
                finishedCallback!!.run()
            }
        }
    }

    val taskCount: Int
        get() {
            synchronized(lock) { return tasks }
        }

    fun getDeferredTasks(): List<DeferredTask> {
        return deferredTasks
    }

    override fun cancelAllDeferred() {}

    override fun <T> flowOnBackground(flow: Flow<T>): Flow<T> {
        return wrappedScheduler.flowOnBackground(flow)
    }

    override fun dispatch(context: CoroutineContext, block: Runnable) {
        increment()
        backgroundDispatcher.dispatch(context) {
            block.run()
            decrement()
        }
    }

    class DeferredTask(
        val tag: String,
        val spec: TaskSpec,
        val repeatPeriod: Long?,
        val inputData: Map<String, String>,
        val networkConstraint: Scheduler.NetworkType?
    )
}
